libr/debug/debug.c

```diff
@@ -505,43 +662,63 @@ R_API int r_debug_step_soft(RDebug *dbg) {
 }
 
 R_API int r_debug_step_hard(RDebug *dbg) {
+	RDebugReasonType reason;
+
 	dbg->reason.type = R_DEBUG_REASON_STEP;
 	if (r_debug_is_dead (dbg)) {
 		return false;
 	}
+
+	/* only handle recoils when not already in recoil mode. */
+	if (dbg->recoil_mode == R_DBG_RECOIL_NONE) {
+		/* handle the stage-2 of breakpoints */
+		if (!r_debug_recoil (dbg, R_DBG_RECOIL_STEP))
+			return false;
+
+		/* recoil already stepped once, so we don't step again. */
+		if (dbg->recoil_mode == R_DBG_RECOIL_STEP) {
+			dbg->recoil_mode = R_DBG_RECOIL_NONE;
+			return true;
+		}
+	}
+
 	if (!dbg->h->step (dbg)) {
 		return false;
 	}
-	return r_debug_wait (dbg);
+	reason = r_debug_wait (dbg);
+	/* TODO: handle better */
+	if (reason == R_DEBUG_REASON_ERROR)
+		return false;
+	if (reason == R_DEBUG_REASON_DEAD || r_debug_is_dead (dbg))
+		return false;
+	return true;
 }
 
 R_API int r_debug_step(RDebug *dbg, int steps) {
```

Comment:
Ok will test this patch on mac and see if it breaks something, if not ill merge. The patch lgtm

> On 21 Jun 2016, at 02:36, Joshua J. Drake <notifications@github.com> wrote:
> 
> In libr/debug/debug.c:
> 
> >  	if (!dbg->h->step (dbg)) {
> >  		return false;
> >  	}
> > -	return r_debug_wait (dbg);
> > +	reason = r_debug_wait (dbg);
> > +	/* TODO: handle better */
> > +	if (reason == R_DEBUG_REASON_ERROR)
> > +		return false;
> > +	if (reason == R_DEBUG_REASON_DEAD || r_debug_is_dead (dbg))
> > +		return false;
> > +	return true;
> >  }
> >  
> >  R_API int r_debug_step(RDebug *dbg, int steps) {
> I have queued up a branch (https://github.com/jduck/radare2/tree/step-return-consistency) for PR following this one.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>